### PR TITLE
Assignment4 - Omar Facundo and Manny Aguilera

### PR DESCRIPTION
**Pull request by omar5o9**
@omar5o9 and @MannyPoptart

### CODE REVIEW

**Comment by EthanAGi** [comment]
I'll review!

File: Logger.Tests/StorageTests.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        Student testStudent = new()
+        {
+            StudentId = 31
+        };
+
+        Assert.Equal(31, testStudent.StudentId);
+    }
+
+
```
Comment (EthanAGi) [inline Logger.Tests/StorageTests.cs:48]: I'd recommend adding more tests, maybe testing for null and empty variables?

File: Logger/BaseClass.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Logger;
+
+public abstract record BaseClass : IEntity
+{
+
+    public Guid Id { get; init; }
+
+    public abstract string Name { set; get; }
```
Comment (EthanAGi) [inline Logger/BaseClass.cs:8]: This is what I'm talking about, I think since you have a setter here the name may be implemented within the interface. I'm unsure on that though.

File: Logger/FullName.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+{
+
+    // We decided to handle the full name storage this way to handle middle names being in the name or not
+    // We are allowing the name to be mutable because names can be changed
+    private string fullName;
+
+    public FullName(string fName, string mName, string lName)
+    {
+
+        fullName = $"{fName} {mName?.Trim()} {lName}";
```
Comment (EthanAGi) [inline Logger/FullName.cs:13]: This is pretty clever! We did something similar, but your code actually looks a lot cleaner than ours!

**Review by EthanAGi — COMMENTED** [review]
This looks really cool! I like the idea of using the person class as an in between for any classes that creates a person! That's very clever!

- Define a full name record (first, last, middle) handling optional and null appropriately. ✔
  - Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
  - Provide a comment on the full name record on why or why not the type is immutable. ✔
- Define an `IEntity` interface: ✔
  - Add an Id property of type `Guid` that is an init-only setter. ✔
  - Add a Name property that is `string`. ✔
- Define an abstract base class that implements `IEntity` - appropriately choosing to implement the interface explicitly or implicitly. ✔
  - Do not implement the `Name` property in this abstract class. ❌ <- I could be incorrect but since you're name variable in the you BaseClass does have a setter, meaning it can be implemented.
  - Do force any derived classes to provide an implementation for `Name`. ✔
- Define book, student, and employee records - all with entity capabilities. ✔
  - Choose a thoughtful approach for the value of `Name` for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
  - Provide a comment on each interface **member** in each entity explaining why you implemented it implicitly or explicitly. ✔
  - You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
  - Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
  - Test that the equality behavior on these entities behaves as expected. ✔
  - Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.
- Unit test the Storage class using various entities to ensure that there are no bugs.
- Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
  - Pay special attention to the CreateLogger methods throughout.

## Fundamentals

- Be sure you enable:
  - Enable Nullability for all projects  ✔
  - Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ✔
  - Ensure that you turn on code analysis for all projects(`EnableNETAnalyzers`) ✔
  - Ensure that you turn on `EnforceCodeStyleInBuild` ✔
  - Ensure that you turn on Nullability (`Nullable`) ✔
- **Ensure there are no errors or warnings (including code analysis warnings)** ✔
- **All of the above should be unit tested** ✔
- **Choose simplicity over complexity** ✔

File: Logger/IEntity.cs [hunk 4]
```diff
@@ -2,6 +2,7 @@
 public interface IEntity
 {
 
-        // Place members here.
+    Guid Id { get; init; }
+    string Name { get; set; }
```
Comment (EthanAGi) [inline Logger/IEntity.cs:6]: Sorry, I placed the comment in the wrong area. This is where I'm a bit worried on the setter, as I believe this may allow the name to be implemented in the interface. I could be incorrect on this.

**Review by rkdeboer — COMMENTED** [review]
Reviewing now

File: Logger/Book.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+namespace Logger;
+
+public record Book : BaseClass
+{
+    public string? Title { get; set; }
+    public string? Author { get; set; }
+    public int YearPublished { get; set; }
+
+    // Calculated property for the full name of the book (Title + Author)
+    public string FullName => $"{Title} by {Author}";
```
Comment (rkdeboer) [inline Logger/Book.cs:11]: I like the stringing together of title and author for full name

File: Logger/Person.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+﻿
+namespace Logger;
+
+public abstract record Person : BaseClass
+{
+    public string? FirstName { get; set; }
```
Comment (rkdeboer) [inline Logger/Person.cs:6]: Could middle name be something to implement?

File: Logger/Student.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿
+
+namespace Logger;
```
Comment (rkdeboer) [inline Logger/Student.cs:3]: Nice work on the person class I did that too!

**Comment by rkdeboer** [comment]
Define a full name record (first, last, middle) handling optional and null appropriately. ✔

Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ✔
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for - - Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
Test that the equality behavior on these entities behaves as expected. ✔

Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.✔
Unit test the Storage class using various entities to ensure that there are no bugs.✔

Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Ensure that you turn on Nullability (Nullable) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: Logger.Tests/StorageTests.cs (hunk 1 above)
Comment (omar5o9) [reply Logger.Tests/StorageTests.cs]: Thanks for the advice, we ended up adding a lot more test to make sure entities work correctly

File: Logger/FullName.cs (hunk 3 above)
Comment (omar5o9) [reply Logger/FullName.cs:13]: Thanks :)

File: Logger/Book.cs (hunk 5 above)
Comment (omar5o9) [reply Logger/Book.cs:11]: Thanks :)

File: Logger/IEntity.cs (hunk 4 above)
Comment (MannyPoptart) [reply Logger/IEntity.cs:6]: Name here should be fine because we made a new name inside of BaseClass so it would not conflict with this one.

File: Logger/Person.cs (hunk 6 above)
Comment (MannyPoptart) [reply Logger/Person.cs:7]: Good catch, we implemented middle name into Person.

File: Logger.Tests/StorageTests.cs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    private readonly Storage storage = new();
+    private readonly Book book = new();
+    private readonly Student student = new();
+    private readonly Employee employee = new();
+    
+    
+
+
+    [Fact]
+    public void Employee_Is_Created_WithProperDepartmentAndID()
```
Comment (BenjaminMichaelis) [inline Logger.Tests/StorageTests.cs:16]:
```suggestion
    public void Employee_Is_CreatedWithProperDepartmentAndID()
```

File: Logger/Book.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+{
+    // Title, Author, and YearPublished are implemented implicitely because they do not
+    // derive from BaseClass
+    public string? Title { get; set; }
+    public string? Author { get; set; }
+    public int YearPublished { get; set; }
+
+    public string FullName => $"{Title} by {Author}";
+
+    public override string Name { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
```
Comment (BenjaminMichaelis) [inline Logger/Book.cs:14]: Why are you throwing a `NotImplementedException`

File: Logger/Employee.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+namespace Logger;
+
+public record Employee : Person
+{
+
+    // EmployeeId and Department are implemented implicitely because they do not
+    // derive from BaseClass
+    public int EmployeeId { get; set; }
+    public string? Department { get; set; }
+    public override string Name { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
```
Comment (BenjaminMichaelis) [inline Logger/Employee.cs:11]: Why are you throwing a `NotImplementedException`

File: Logger/FullName.cs [hunk 11]
```diff
@@ -0,0 +1,16 @@
+﻿namespace Logger;
+
+public record FullName
+{
+
+    // We decided to handle the full name storage this way to handle middle names being in the name or not
+    // We are allowing the name to be mutable because names can be changed
+    private string fullName;
+
+    public FullName(string fName, string mName, string lName)
```
Comment (BenjaminMichaelis) [inline Logger/FullName.cs:10]: we don't have to worry about longer names having some performance boost or something. full names are more clear.\
```suggestion
    public FullName(string firstName, string middleName, string lastName)
```

File: Logger/IEntity.cs [hunk 12]
```diff
@@ -2,6 +2,7 @@
 public interface IEntity
 {
 
-        // Place members here.
+    Guid Id { get; init; }
```
Comment (BenjaminMichaelis) [inline Logger/IEntity.cs:5]: Why do we need `init`?

File: Logger/Student.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+public record Student : Person
+{
+    // StudentId is implement implicitely because it does not derive from BaseClass
+    public int StudentId { get; set; }
+    public override string Name { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
```
Comment (BenjaminMichaelis) [inline Logger/Student.cs:9]: Why the `NotImplementedException`

File: Logger.Tests/StorageTests.cs
Comment (BenjaminMichaelis) [inline Logger.Tests/StorageTests.cs:1]: There are lots of tests that are missing for each model type
